src/main/java/io/cdap/plugin/cdc/source/salesforce/SalesforceReceiver.java

```diff
@@ -0,0 +1,250 @@
+                     ErrorHandling errorHandling) {
+    super(StorageLevel.MEMORY_AND_DISK_2());
+    this.credentials = credentials;
+    this.objectsForTracking = new ArrayList<>(objectsForTracking);
+    this.errorHandling = errorHandling;
+  }
+
+  @Override
+  public void onStart() {
+    eventTopicListener = new SalesforceEventTopicListener(credentials, objectsForTracking);
+    eventTopicListener.start();
+
+    ThreadFactory namedThreadFactory = new ThreadFactoryBuilder()
+      .setNameFormat(RECEIVER_THREAD_NAME + "-%d")
+      .build();
+
+    Executors.newSingleThreadExecutor(namedThreadFactory).submit(this::receive);
+  }
+
+  @Override
+  public void onStop() {
+    // There is nothing we can do here as the thread calling receive()
+    // is designed to stop by itself if isStopped() returns false
+  }
+
+  private void receive() {
+    PartnerConnection connection;
+    try {
+      connection = SalesforceConnectionUtil.getPartnerConnection(credentials);
+    } catch (ConnectionException e) {
+      throw new RuntimeException("Failed to connect to Salesforce", e);
+    }
+
+    while (!isStopped()) {
+      try {
+        String message = eventTopicListener.getMessage(GET_MESSAGE_TIMEOUT_SECONDS, TimeUnit.SECONDS);
+
+        if (message != null) {
+          // whole message class is not needed because we are interested only in change event payload
+          JsonObject headerElement = JSON_PARSER.parse(message)
+            .getAsJsonObject()
+            .getAsJsonObject("data")
+            .getAsJsonObject("payload")
+            .getAsJsonObject("ChangeEventHeader");
+          ChangeEventHeader event = GSON.fromJson(headerElement, ChangeEventHeader.class);
+
+          List<ChangeEventHeader> eventsList = events.getOrDefault(event.getTransactionKey(), new ArrayList<>());
+          eventsList.add(event);
+
+          if (event.isTransactionEnd()) {
+            processEvents(eventsList, connection);
+            events.remove(event.getTransactionKey());
+          } else {
+            events.put(event.getTransactionKey(), eventsList);
+          }
+        }
+      } catch (Exception e) {
+        switch (errorHandling) {
+          case SKIP:
+            LOG.warn("Failed to process message, skipping it.", e);
+            break;
+          case STOP:
+            throw new RuntimeException("Failed to process message", e);
+          default:
+            throw new IllegalStateException(String.format("Unknown error handling strategy '%s'", errorHandling));
+        }
+      }
+    }
+    eventTopicListener.stop();
+  }
+
+  private void processEvents(List<ChangeEventHeader> events, PartnerConnection connection) throws ConnectionException {
+    for (ChangeEventHeader event : events) {
+      SObjectDescriptor descriptor = SObjectDescriptor.fromName(event.getEntityName(), connection);
+      SObjectsDescribeResult describeResult = new SObjectsDescribeResult(connection, descriptor.getAllParentObjects());
```

Comment:
is this looking up the SObject descriptor as it currently is? Does the change event not have all the information required to apply the change? If not, I don't understand how you can apply the right change (or how you can do anything efficiently), as this object may be different now than it was when the change event occurred. Unless objects cannot change?

I thought I asked these questions in the design doc. Were they ever addressed?

Reply:
For CREATE, DELETE, UNDELETE, UPDATE types change event message contains information only about fields which were changed. If type of event is GAP_CREATE, GAP_DELETE, GAP_UNDELETE, GAP_UPDATE, GAP_OVERFLOW than it will contain only ids of records. So we have to make requests to salesforce to get full information.